clipdel

```diff
@@ -16,72 +16,164 @@ cache_file_prefix=$cache_dir/line_cache
 lock_file=$cache_dir/lock
 lock_timeout=2
 
-if [[ $1 == --help ]] || [[ $1 == -h ]]; then
-    cat << 'EOF'
-clipdel deletes clipmenu entries matching a regex. By default, just lists what
-it would delete, pass -d to do it for real.
+line_cache_files=( "$cache_file_prefix"_* )
+
+check_conflict() {
+    if [[ -z "$CM_FUNCTION" ]]; then
+        CM_FUNCTION="$1"
+        CM_REQUIRE_PATTERN=$2
+        if ! ((CM_REQUIRE_PATTERN)) && ((${#CM_PATTERNS[@]})); then
+            printf '%s\n' "Too many arguments" >&2
+            return 2
+        fi
+    else
+       printf '%s\n' "${3:-Options '$1' and '$CM_FUNCTION' conflict}" >&2
+       return 1
+    fi
+}
 
-".*" is special, it will just nuke the entire data directory, including the
-line caches and all other state.
+print_usage() {
+    cat << 'EOF'
+Usage: clipdel [OPTION] [REGEX]
+Delete clipmenu entries.
 
 Arguments:
 
-    -d  Delete for real.
+    --all                      Delete all entries
+    -c, --current              Delete current entry
+    -d, --delete               Delete for real (default when deleting all, current or from stdin)
+    -t, --older-than <STRING>  Delete entries older than STRING (format: N UNIT [N UNIT...])
+    -h, --help                 Print this message
+
+Examples:
+
+    clipmenu | clipdel         Launch clipmenu and delete selected entry
+    clipdel -d "bye world"     Delete all entries containing "bye world"
+    clipdel -t "10 hours" -d   Delete all entries older than 10 hours
+    clipdel -t "1 year 1 day"  List entries older than 1 year plus 1 day
 
 Environment variables:
 
 - $CM_DIR: specify the base directory to store the cache dir in (default: $XDG_RUNTIME_DIR, $TMPDIR, or /tmp)
 EOF
-    exit 0
-fi
-
-line_cache_files=( "$cache_file_prefix"_* )
+}
+
+while (($# > 0)); do
+    case "$1" in
+        --all|-c|--current) check_conflict "$1" 0 || exit 1;;
+        -d|--delete) CM_REAL_DELETE=1;;
+        -h|--help)
+            print_usage
+            exit 0
+            ;;
+        -t|--older-than)
+            check_conflict "$1" 0 || exit 1
+            shift
+            if [[ "$1" ]]; then
+                date_format=$(sed "s,\([0-9]\+ [A-Za-z]\+\),\1 ago,g" <<< "$1")
+                CM_OLDER_THAN=$(date --date="$date_format" +%s%N 2> /dev/null) ||
+                {
+                    printf '%s\n' "Unrecognized date format '$1'" >&2
+                    exit 1
+                }
+            else
+                printf '%s\n' "Missing date format" >&2
+                exit 1
+            fi
+            ;;
+        --|[!-]*)
+            if ((CM_REQUIRE_PATTERN)) && ! ((${#CM_PATTERNS[@]})); then
```

Comment:
I find this logic a bit hard to reason about. Can it be implemented with slicing and a for loop instead? :-)

Reply:
This is so everything after a `--` is read as a pattern. E.g.:

`clipdel --all`: will delete all entries
`clipdel -- --all`: will delete all entries matching `--all`

`clipdel pattern1 -- pattern2` will fail due to excessive arguments.

Sorry, I don't understand what you mean by "be implemented with slicing and a for loop instead". Could you give me a quick draft?

Reply:
Right, I'm talking about something like this: https://github.com/cdown/mpdmenu/blob/master/mpdmenu#L36-L40